### PR TITLE
Fix accessing symbols introduced within immediate from clause, through the frame symbol

### LINKED ISSUE #32951: Query expression inside a query action gives a compile error
Consider the following code.

```ballerina
public function main () returns error?{
    int[][] data = [[1,2], [2,3,4]];
    int sumofEven = 0;
    check from int[] arr in data do { // Query Action
        int[] evenNumbers = from int i in arr where i % 0 == 0 select i; // Query expression
        check from int i in evenNumbers do {
            sumofEven += i;
        };
    };
}
```

Compiling the above code gives an error.

```
incompatible types: expected 'other', found 'int'
undefined symbol 'evenNumbers'
```

### LINKED ISSUE #33709: Query expression result in an unexpected runtime error
**Description:**
```ballerina
type BddPath record {|
    int[] pos = [1];
|};

public type MappingAlternative record {|
    int[] pos;
|};

public function main() {
    BddPath[] paths = [{}, {}, {}];

    MappingAlternative[] alts = from var {pos} in paths
        select {
            pos: (from var atom in pos
                select atom)
        };
}
```

```

error: {ballerina}TypeCastError {"message":"incompatible types: '()' cannot be cast to 'int'"}
        at ballerina.lang.query.0._SelectFunction:process(types.bal:590)
           ballerina.lang.query.0:$annon$method$delegate$0(types.bal:91)
           ballerina.lang.__internal.0:invokeAsExternal(stream-utils.bal:138)
           ballerina.lang.query.0._StreamPipeline:next(types.bal:91)
           ballerina.lang.query.0.IterHelper:next(types.bal:723)
           ballerina.lang.stream.0:next(stream.bal:67)
           ballerina.lang.query.0:toArray(helpers.bal:88)
           main:main(main.bal:12)
```

**Steps to reproduce:**

**Affected Versions:**
Beta-3
**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #33709 
Fixes #32951 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal [hunk 1]
```diff
@@ -302,6 +302,57 @@ function testTypeTestInWhereClause() {
...
+                int[] evenNumbers = from int i in arr
+                    where i % 2 == 0
+                    select i;
+                return evenNumbers;
+            };
+            int[] expected = [2, 4];
+            assertEquality(expected, func());
+        };
+}
+
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal:355]: Shouldn't the following also work?
```ballerina
type A record {|
    int[] pos;
|};

public function main() {
    A[] res = from var a in 2...5
        select {
            pos: (from var b in (from var c in 1...2 where c == a select c) // undefined symbol 'a'
                select b)
        };
}
```

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal:355]: Here the issue is of two parts.
1. Parent env is set incorrectly (fixed with 8c79123497c8c7eed5bcf4c9b8536db9966b23b1)
2. Closure issue in nested queries (https://github.com/ballerina-platform/ballerina-lang/issues/32710) which requires QueryDesugar rewrite

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal:355]: Shall we put this example in #32710 to track it and address it in that issue?

**Review by gimantha — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal (hunk 1 above)
Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/inner-queries.bal:355]: +1

**Review by lasinicl — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 2]
```diff
@@ -1500,7 +1501,8 @@ public void visit(BLangSimpleVarRef bLangSimpleVarRef) {
         // check whether the symbol and resolved symbol are the same.
         // because, lookup using name produce unexpected results if there's variable shadowing.
         if (symbol != null && symbol != resolvedSymbol && !FRAME_PARAMETER_NAME.equals(identifier)) {
-            if (!identifiers.containsKey(identifier)) {
+            if (!identifiers.containsKey(identifier) && (withinLambdaFunc || queryEnv == null
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:1504]: Shall we add `map.containsKey` as second condition in the `&&` that we we wont access map always

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 3]
```diff
@@ -1997,6 +2003,7 @@ public void visit(BLangForeach foreach) {
 
     @Override
     public void visit(BLangFromClause fromClause) {
+        this.queryEnv = fromClause.env;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:2006]: Shall we add a comment to explain why we dont have to restore the env to `prevQueryEnv` here

**Review by KRVPerera — COMMENTED** [review]
Add some comment in-case we need to merge this PR

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 3 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:2006]: Added in b7f8885

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 2 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: Addressed in https://github.com/ballerina-platform/ballerina-lang/commit/b7f888577eff0080b9de31022179665af240b4e6

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -5484,13 +5479,19 @@ public void visit(BLangQueryAction queryAction) {
...
+        if (!breakToParallelQueryEnv) {
+            prevEnvs.pop();
+        }
     }
 
     @Override
     public void visit(BLangFromClause fromClause) {
         boolean prevBreakToParallelEnv = this.breakToParallelQueryEnv;
-        this.breakToParallelQueryEnv = true;
+        if (fromClause.collection.getKind() == NodeKind.QUERY_EXPR ||
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5490]: shall we introduce a variable here for `fromClause.collection`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 5]
```diff
@@ -5504,7 +5505,11 @@ public void visit(BLangFromClause fromClause) {
     @Override
     public void visit(BLangJoinClause joinClause) {
         boolean prevBreakEnv = this.breakToParallelQueryEnv;
-        this.breakToParallelQueryEnv = true;
+        if (joinClause.collection.getKind() == NodeKind.QUERY_EXPR ||
+                (joinClause.collection.getKind() == NodeKind.GROUP_EXPR
+                        && ((BLangGroupExpr) joinClause.collection).expression.getKind() == NodeKind.QUERY_EXPR)) {
+            this.breakToParallelQueryEnv = true;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5511]: Shall we explain what is parallel query expression breaking near this variables definition with a comment.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -5484,13 +5479,19 @@ public void visit(BLangQueryAction queryAction) {
...
     }
 
     @Override
     public void visit(BLangFromClause fromClause) {
         boolean prevBreakToParallelEnv = this.breakToParallelQueryEnv;
-        this.breakToParallelQueryEnv = true;
+        if (fromClause.collection.getKind() == NodeKind.QUERY_EXPR ||
+                (fromClause.collection.getKind() == NodeKind.GROUP_EXPR
+                        && ((BLangGroupExpr) fromClause.collection).expression.getKind() == NodeKind.QUERY_EXPR)) {
+            this.breakToParallelQueryEnv = true;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5493]: Some places we have used `this.breakToParallelQueryEnv` but in other occasions we don't use `this`. Shall we make the code consistent in all the usages of this variable.

**Review by lasinicl — APPROVED** [review]

**Review by gimantha — APPROVED** [review]

**Review by gimantha — APPROVED** [review]

**Review by lasinicl — APPROVED** [review]

**Pull request merged**